### DISCUSSION TITLE
Bench testing Inputs advices

### DISCUSSION THREAD

**Discussion by bzhmaddog** [bitfieldlabs/afterglow#15] [General]
Hello there.
I have assembled my first Afterglow board (v1.3) with intent to use on my f-14 Tomcat machine. I have successfully tested the outputs using the test mode (dip 1 set to on) and all the slots of the matrix works fine. I was wondering what is the best way to test the input wihout having to plug it in the machine.

I am afraid to inject the 18v directly on col/row (I'm not clear on the polarity)

Any advice appreciated

**Comment by bzhmaddog** [comment]
For the context, I have tested the board in my machine and I think there is something wrong (but no smoke so far).
All lamps stays on (every colums and rows).
Is there a configuration that I missed ? (i haven't used the configuration tool yet)

**Comment by bzhmaddog** [comment]
After a board inspection I found some suspicious connections on U2 so I resoldered the component and now none of the lamps lights up.
Still not working though, now when I apply voltage to any of the Columns the red led on the Nano stops blinking (stays on). I am not sure if it is intented.

**Comment by bitfieldlabs** [comment]
For testing the inputs outside of a pinball machine you need a test setup. I usually do it with a second afterglow board, set to test and pass-through mode. But testing in the game is not really risky. At least not for the pinball machine ;-)

The most important thing is to never hook the AG board to the pinball machine without connecting ground! Doing so will kill your input shift registers (U2). **Never apply voltage to the columns without ground connected.**

If the input columns and rows are read correctly, the status LED on PCB v1.3-1.5 should stay on permanently. On error it will stay either off or starts blinking quickly.

For debugging the inputs you can compile the Arduino firmware with `DEBUG_SERIAL` set to 1 and read the debug information from the USB serial monitor at 115200 baud.

**Comment by bzhmaddog** [comment]
I did compile the firmware with debug mode on and U2's ground pin is OK.
When just plugin the AG board in my computer (without plugin the I/O) I get theses 'logs'

```
afterglow v110
-------------------------------
(c) 2018-2022 morbid cornflakes
efuse BOD 2.7V
CFG from EEPROM
TTAG_INT 150
INT dt max 123us last 119us ovfl 0
Bad col: 30078 col 0 ord 0 last good: 0
IN C 0x0000 R 0xFF00
OUT C 0x0008 R 0x0000
C0 + 0 0 0 0 0 0 0 0 
C1 + 0 0 0 0 0 0 0 0 
C2 + 0 0 0 0 0 0 0 0 
C3 + 0 0 0 0 0 0 0 0 
C4 + 0 0 0 0 0 0 0 0 
C5 + 0 0 0 0 0 0 0 0 
C6 + 0 0 0 0 0 0 0 0 
C7 + 0 0 0 0 0 0 0 0 
TTAG_INT 150
INT dt max 123us last 121us ovfl 0
Bad col: 63830 col 0 ord 0 last good: 0
IN C 0x0000 R 0xFF00
OUT C 0x0008 R 0x0000
C0 + 0 0 0 0 0 0 0 0 
C1 + 0 0 0 0 0 0 0 0 
C2 + 0 0 0 0 0 0 0 0 
C3 + 0 0 0 0 0 0 0 0 
C4 + 0 0 0 0 0 0 0 0 
C5 + 0 0 0 0 0 0 0 0 
C6 + 0 0 0 0 0 0 0 0 
C7 + 0 0 0 0 0 0 0 0 
TTAG_INT 150
INT dt max 123us last 119us ovfl 0
Bad col: 97585 col 0 ord 0 last good: 0
IN C 0x0000 R 0xFF00
OUT C 0x0008 R 0x0000
C0 + 0 0 0 0 0 0 0 0 
C1 + 0 0 0 0 0 0 0 0 
C2 + 0 0 0 0 0 0 0 0 
C3 + 0 0 0 0 0 0 0 0 
C4 + 0 0 0 0 0 0 0 0 
C5 + 0 0 0 0 0 0 0 0 
C6 + 0 0 0 0 0 0 0 0 
C7 + 0 0 0 0 0 0 0 0 
TTAG_INT 150
INT dt max 123us last 121us ovfl 0
Bad col: 131334 col 0 ord 0 last good: 0
IN C 0x0000 R 0xFF00
OUT C 0x0010 R 0x0000
C0 + 0 0 0 0 0 0 0 0 
C1 + 0 0 0 0 0 0 0 0 
C2 + 0 0 0 0 0 0 0 0 
C3 + 0 0 0 0 0 0 0 0 
C4 + 0 0 0 0 0 0 0 0 
C5 + 0 0 0 0 0 0 0 0 
C6 + 0 0 0 0 0 0 0 0 
C7 + 0 0 0 0 0 0 0 0 
TTAG_INT 150
INT dt max 123us last 119us ovfl 0
Bad col: 165088 col 0 ord 0 last good: 0
IN C 0x0000 R 0xFF00
OUT C 0x0040 R 0x0000
C0 + 0 0 0 0 0 0 0 0 
C1 + 0 0 0 0 0 0 0 0 
C2 + 0 0 0 0 0 0 0 0 
C3 + 0 0 0 0 0 0 0 0 
C4 + 0 0 0 0 0 0 0 0 
C5 + 0 0 0 0 0 0 0 0 
C6 + 0 0 0 0 0 0 0 0 
C7 + 0 0 0 0 0 0 0 0 
```

I understand the meaning of C0 + xxxxxx to C7 + xxxx but can you explain the other lines ?
Does 'Bad col' means something is wrong ?
Same for IN C 0x0000 R 0xFF00 <==== I would expect R to be 0x0000 too right ?

I'll debug in the machine to try to find differences

**Reply by bitfieldlabs** [reply]
From a correct column input the firmware expects always one single column to be active (+18V). Everything which has either none or multiple lines at +18V is considered a bad column input. So it's normal that this value counts up when no column input is connected.

The "Int dt max" line just indicates how long the code executed in the last timer interrupt (executed at 4kHz).

**Reply by bitfieldlabs** [reply]
As the rows are pulled high by default, FF means all rows off.

**Reply by bzhmaddog** [reply]
Well it seems it actually works just fine, there might have been something wrong with my wiring harness when I tested Yesterday.
Now I just need to wait for the adapter PCB I made to be able to test with the playfield.

Thx for the help.

Logs from the All Lamps Test
------------------------------------------

10:45:00.658 -> afterglow v110
10:45:00.658 -> -------------------------------
10:45:00.658 -> (c) 2018-2022 morbid cornflakes
10:45:00.658 -> efuse BOD 2.7V
10:45:00.658 -> CFG from EEPROM
10:45:05.147 -> TTAG_INT 150
10:45:05.147 -> INT dt max 142us last 119us ovfl 0
10:45:05.202 -> Bad col: 284 col 12 ord 1 last good: 0
10:45:05.202 -> IN C 0x0001 R 0xFFFF
10:45:05.202 -> OUT C 0x0008 R 0x00FF
10:45:05.202 -> C0 + 59904 59904 59904 65535 65535 65535 65535 65535 
10:45:05.212 -> C1 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:05.212 -> C2 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:05.244 -> C3 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:05.244 -> C4 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:05.277 -> C5 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:05.277 -> C6 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:05.317 -> C7 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:10.325 -> TTAG_INT 150
10:45:10.325 -> INT dt max 142us last 117us ovfl 0
10:45:10.325 -> Bad col: 610 col 24 ord 1 last good: 3
10:45:10.325 -> IN C 0x0008 R 0xFFFF
10:45:10.325 -> OUT C 0x0040 R 0x00FF
10:45:10.325 -> C0 + 44928 44928 44928 44928 44928 44928 44928 44928 
10:45:10.378 -> C1 + 52416 52416 52416 52416 52416 52416 52416 52416 
10:45:10.378 -> C2 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:10.390 -> C3 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:10.423 -> C4 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:10.423 -> C5 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:10.455 -> C6 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:10.455 -> C7 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:15.470 -> TTAG_INT 150
10:45:15.470 -> INT dt max 142us last 118us ovfl 0
10:45:15.470 -> Bad col: 938 col 48 ord 1 last good: 6
10:45:15.521 -> IN C 0x0080 R 0xFFFF
10:45:15.521 -> OUT C 0x0080 R 0x0000
10:45:15.521 -> C0 + 22464 22464 22464 22464 22464 29952 29952 29952 
10:45:15.521 -> C1 + 29952 29952 29952 29952 29952 37440 37440 37440 
10:45:15.535 -> C2 + 37440 37440 37440 37440 44928 44928 44928 44928 
10:45:15.535 -> C3 + 52416 52416 52416 52416 59904 59904 59904 59904 
10:45:15.568 -> C4 + 59904 59904 59904 59904 65535 65535 65535 65535 
10:45:15.568 -> C5 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:15.600 -> C6 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:15.632 -> C7 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:20.647 -> TTAG_INT 150
10:45:20.647 -> INT dt max 142us last 117us ovfl 0
10:45:20.647 -> Bad col: 1270 col 3 ord 1 last good: 1
10:45:20.647 -> IN C 0x0004 R 0xFF00
10:45:20.647 -> OUT C 0x0002 R 0x0000
10:45:20.647 -> C0 + 0 0 0 0 0 0 0 0 
10:45:20.647 -> C1 + 0 0 7488 7488 7488 7488 7488 7488 
10:45:20.693 -> C2 + 7488 7488 7488 7488 7488 14976 14976 14976 
10:45:20.693 -> C3 + 14976 14976 14976 14976 14976 22464 22464 22464 
10:45:20.714 -> C4 + 29952 29952 29952 29952 29952 37440 37440 37440 
10:45:20.714 -> C5 + 37440 37440 37440 37440 44928 44928 44928 44928 
10:45:20.747 -> C6 + 44928 44928 44928 44928 52416 52416 52416 52416 
10:45:20.747 -> C7 + 52416 52416 52416 52416 59904 59904 59904 59904 
10:45:25.782 -> TTAG_INT 150
10:45:25.782 -> INT dt max 142us last 115us ovfl 0
10:45:25.782 -> Bad col: 1586 col 129 ord 1 last good: 0
10:45:25.782 -> IN C 0x0001 R 0xFF00
10:45:25.782 -> OUT C 0x0008 R 0x0000
10:45:25.782 -> C0 + 0 0 0 0 0 0 0 0 
10:45:25.836 -> C1 + 0 0 0 0 0 0 0 0 
10:45:25.836 -> C2 + 0 0 0 0 0 0 0 0 
10:45:25.836 -> C3 + 0 0 0 0 0 0 0 0 
10:45:25.836 -> C4 + 0 0 0 0 0 0 0 0 
10:45:25.836 -> C5 + 0 0 0 0 0 0 0 0 
10:45:25.836 -> C6 + 0 0 0 0 0 0 0 0 
10:45:25.836 -> C7 + 0 0 0 0 0 0 0 0 
10:45:30.841 -> TTAG_INT 150
10:45:30.841 -> INT dt max 142us last 117us ovfl 0
10:45:30.841 -> Bad col: 1912 col 48 ord 1 last good: 4
10:45:30.841 -> IN C 0x0020 R 0xFF00
10:45:30.841 -> OUT C 0x0010 R 0x0000
10:45:30.841 -> C0 + 0 0 0 0 0 0 0 0 
10:45:30.841 -> C1 + 0 0 0 0 0 0 0 0 
10:45:30.841 -> C2 + 0 0 0 0 0 0 0 0 
10:45:30.892 -> C3 + 0 0 0 0 0 0 0 0 
10:45:30.892 -> C4 + 0 0 0 0 0 0 0 0 
10:45:30.892 -> C5 + 0 0 0 0 0 0 0 0 
10:45:30.892 -> C6 + 0 0 0 0 0 0 0 0 
10:45:30.892 -> C7 + 0 0 0 0 0 0 0 0
10:45:35.877 -> TTAG_INT 150
10:45:35.877 -> INT dt max 142us last 117us ovfl 0
10:45:35.877 -> Bad col: 2226 col 6 ord 1 last good: 1
10:45:35.933 -> IN C 0x0002 R 0xFF00
10:45:35.933 -> OUT C 0x0080 R 0x0000
10:45:35.933 -> C0 + 0 0 0 0 0 0 0 0 
10:45:35.933 -> C1 + 0 0 0 0 0 0 0 0 
10:45:35.933 -> C2 + 0 0 0 0 0 0 0 0 
10:45:35.933 -> C3 + 0 0 0 0 0 0 0 0 
10:45:35.933 -> C4 + 0 0 0 0 0 0 0 0 
10:45:35.953 -> C5 + 0 0 0 0 0 0 0 0 
10:45:35.953 -> C6 + 0 0 0 0 0 0 0 0 
10:45:35.953 -> C7 + 0 0 0 0 0 0 0 0 
10:45:40.934 -> TTAG_INT 150
10:45:40.934 -> INT dt max 142us last 118us ovfl 0
10:45:40.991 -> Bad col: 2543 col 192 ord 1 last good: 6
10:45:40.991 -> IN C 0x00C0 R 0xFFFF
10:45:40.991 -> OUT C 0x0004 R 0x00FF
10:45:40.991 -> C0 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:40.991 -> C1 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:41.000 -> C2 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:41.042 -> C3 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:41.042 -> C4 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:41.042 -> C5 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:41.121 -> C6 + 65535 65535 65535 0 0 0 0 0 
10:45:41.121 -> C7 + 0 0 0 0 0 0 0 0
10:45:46.094 -> TTAG_INT 150
10:45:46.094 -> INT dt max 142us last 118us ovfl 0
10:45:46.094 -> Bad col: 2867 col 3 ord 1 last good: 7
10:45:46.094 -> IN C 0x0080 R 0xFFFF
10:45:46.094 -> OUT C 0x0020 R 0x00FF
10:45:46.094 -> C0 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:46.152 -> C1 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:46.152 -> C2 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:46.162 -> C3 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:46.162 -> C4 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:46.194 -> C5 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:46.194 -> C6 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:46.234 -> C7 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:51.247 -> TTAG_INT 150
10:45:51.247 -> INT dt max 142us last 120us ovfl 0
10:45:51.247 -> Bad col: 3194 col 12 ord 1 last good: 3
10:45:51.247 -> IN C 0x0008 R 0xFFFF
10:45:51.247 -> OUT C 0x0004 R 0x00FF
10:45:51.247 -> C0 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:51.306 -> C1 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:51.306 -> C2 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:51.315 -> C3 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:51.347 -> C4 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:51.347 -> C5 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:51.379 -> C6 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:51.379 -> C7 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:56.389 -> TTAG_INT 150
10:45:56.389 -> INT dt max 142us last 118us ovfl 0
10:45:56.444 -> Bad col: 3512 col 48 ord 1 last good: 7
10:45:56.444 -> IN C 0x0080 R 0xFFFF
10:45:56.444 -> OUT C 0x0080 R 0x00FF
10:45:56.444 -> C0 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:56.455 -> C1 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:56.455 -> C2 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:56.488 -> C3 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:56.488 -> C4 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:56.520 -> C5 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:56.520 -> C6 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:45:56.566 -> C7 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:46:01.578 -> TTAG_INT 150
10:46:01.578 -> INT dt max 142us last 118us ovfl 0
10:46:01.578 -> Bad col: 3852 col 6 ord 1 last good: 2
10:46:01.578 -> IN C 0x0008 R 0xFFFF
10:46:01.578 -> OUT C 0x0040 R 0x00FF
10:46:01.578 -> C0 + 44928 52416 52416 52416 52416 52416 52416 52416 
10:46:01.639 -> C1 + 52416 59904 59904 59904 59904 59904 59904 59904 
10:46:01.639 -> C2 + 59904 65535 65535 65535 65535 65535 65535 65535 
10:46:01.647 -> C3 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:46:01.648 -> C4 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:46:01.677 -> C5 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:46:01.677 -> C6 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:46:01.712 -> C7 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:46:06.725 -> TTAG_INT 150
10:46:06.725 -> INT dt max 142us last 118us ovfl 0
10:46:06.725 -> Bad col: 4192 col 24 ord 1 last good: 6
10:46:06.725 -> IN C 0x0040 R 0xFFFF
10:46:06.785 -> OUT C 0x0040 R 0x0000
10:46:06.785 -> C0 + 29952 29952 29952 29952 29952 37440 37440 37440 
10:46:06.785 -> C1 + 37440 37440 37440 37440 37440 44928 44928 44928 
10:46:06.785 -> C2 + 44928 44928 44928 44928 44928 52416 52416 52416 
10:46:06.796 -> C3 + 59904 59904 59904 59904 65535 65535 65535 65535 
10:46:06.827 -> C4 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:46:06.827 -> C5 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:46:06.860 -> C6 + 65535 65535 65535 65535 65535 65535 65535 65535 
10:46:06.860 -> C7 + 65535 65535 65535 65535 65535 65535 65535 65535

**Comment by bzhmaddog** [comment]
Ok so I have finally been able to test with the playfield and unfortunately it is not working as expected.
I have three different behaviors (maybe all linked).
1 - When the machine is in attract mode The light show does not look like normal (probably due to some ghosting).
2 - In the 'All Lamps' test mode none of the lamps on row 3 are working/blinking.
3 - When in the 'Individual lamp' test mode, each lamp lights more lamps than it should (Ghosting ?).

Continuity of Row 3 seems ok on both side of the board.

Any pointer to where to look for troubleshooting ?

**Comment by bitfieldlabs** [comment]
What happens in AG test mode?
As stated before, the AG board itself cannot really cause ghosting. If you encounter it there is either something wrong with your playfield lamp matrix or the AG board is defective.
You can also try and see what happens in pass-through mode (switch 4).

**Comment by bzhmaddog** [comment]
In test mode I have the same 'ghosting' effect. and In Passthrough it is also not good (all lamps stays on almost all the time and it looks like the attract mode is playing at high speed).
I don't think the problem lies within the playfield lamp matrix. I never noticed any ghosting while playing on in test mode.
I probably messed up something during the build but I am not sure where to start trouble shooting.
I wish the Test mode would alllow to cycle through columns/rows manually (using a push button) like the manual lamp test of the machine.
I guess I could tweak the code and use one of the free i/o pins of the nano to do that.
I'm also building a 8x8 led matrix with a 1088AS matrix to be able to bench test.

**Reply by bitfieldlabs** [reply]
It's difficult to debug it remotely. It would be much easier if you send the board to me. There must be something wrong on the board itself.

**Reply by bzhmaddog** [reply]
Thx for the offer but I am not gonna bother you with that. I will build another board with spare parts I have.
I build myself a lamp matrix tester and recorded a video of the test patterns the board is outputing

I uploaded it to youtube => https://www.youtube.com/watch?v=TljudarRksI

Does it look right to you ? at 0.41 the pattern looks a bit wonky. Not all leds are 'glowing' Is it in purpose ?

**Reply by bitfieldlabs** [reply]
Yes, that looks fine to me. Not sure what happens at 0:41. Maybe I screwed up the pattern, the LEDs are all fine in the other loops.

**Comment by bzhmaddog** [comment]
Well. I think I fried my board but I am not sure how.
All of a sudden the led on the arduino stopped blinking (blink once then stays off).
It looks like I have a short somewhere, my bench power supply drop the 18V output to 3.7V and the Current control led turns on.
When I remove the diode (D1) from the circuit the Powe supply doesn't do that and I get correct 18V at the mosfets.
My guess is that one of the 5V powered (U1,U2,U5,U6) is dead short (The arduino works fine out of the circuit).

**Comment by bitfieldlabs** [comment]
Oh, too bad. It's typically U2 which fries whenever there is a problem with grounding.

**Comment by bzhmaddog** [comment]
Indeed U2 was dead. I replaced it and now the short is gone but I have all leds in one column not working. Could the associated mosfet (Q4) have been damaged ? or U7 ?
